README.md
![logo](https://s3.amazonaws.com/f.cl.ly/items/1u233B1J332D393D2m2j/avatar-logo.png)
================================================================================

Consolidated Avatar Template Package for Meteor
-----------------------------------------------

ANNOUNCEMENT
------------

This package has been forked from @bengott's original Avatar package since it is longer maintained. It's been added to the *utilities* collection so that it can henceforth be maintained by other community members.

BREAKING CHANGES
----------------

As of version 0.9.0 the React support that was added to version 0.8.0 has been removed. The latest updates to Meteor and the React package have made the build unstable, and despite my best efforts I haven't been able to create a single solution that works with both apps that use and don't use React. Days have been lost trying to solve the issues, and this has stopped me from publishing other updates. This situation will be revisited at a later stage, or else a separate React fork may be considered. In the meantime you'll need to use the Blaze template within your React code.

In version 0.7.14 and earlier the Avatar options were overridden by assigning them directly to `Avatar.options`. You should no longer do this. Instead, you must call `Avatar.setOptions()` and pass the options that you wish to override.

The template parameters were overhauled in version 0.5.0. The `Avatar.options` object changed quite a bit in version 0.6.0 too. And make sure you note the `defaultType`/`fallbackType` changes in version 0.7.0. Basically, things are still in a state of flux (pre-1.0.0), so check for breaking changes and read the rest of the README carefully when you update the package.

Installation
------------
In your Meteor project directory, run:
```shell
$ meteor add utilities:avatar
```
Of course, you also need to add the accounts-<service> packages for the services you're using (e.g. accounts-twitter) and accounts-ui or something similar in order to add login functionality to your app.

Usage
-----
In an HTML file:
```handlebars
{{> avatar (user=<user> || userId=<userId>)
           (size="large" || "small" || "extra-small" || <user-defined size>)
           (shape="rounded" || "circle")
           (class="some custom classes")
           (initials="<initials>") (bgColor="<color>") (txtColor="<color>") }}
```

That may look like a lot of options, but they are all optional. Most of the time, your HTML will look more like this:
```handlebars
{{> avatar user=this shape="circle"}}
```

Optional template parameters:
  - `user` or `userId`: Either a user object or userId string, default avatar if omitted
  - `size`: Size of the avatar, by default either "large" (80px), "small" (30px), or "extra-small" (20px), although these may be overridden and new ones defined. If no size is provided 50px is used.
  - `shape`: Used for CSS border-radius property, either "rounded" or "circle", square if omitted
  - `class`: Any custom CSS classes you'd like to define on the avatar container. The string is passed straight through to the `class` attribute on the `div` container element.
  - `initials`: Specify the initials to show for the initials avatar. The package automatically tries to determine the user's initials from profile data, but if defined, this param will override that.
  - `bgColor` and `txtColor`: Override the default colors for the initials avatar (color name or hex value string).

Publishing & Subscribing
------------------------

This package will help you display an avatar based on the data available in the user object, but it won't actually make that data available for you. In other words, you still need to publish the relevant user profile fields, and subscribe to that publication yourself somewhere in your app's codebase.

Generated CSS
-------------

The CSS required by this package is generated based on the options that you configure (see below) and then pushed into the `head` of your HTML at the point that the application is served to the client. This approach gives us great flexibility and allows us to provide features that would not otherwise be possible (such as custom avatar sizes).

However, since the CSS isn't generated when you build the project, you won't have any styling for Cordova apps (until they talk to the server). In this case you'll need to provide your own CSS for the avatars.

If you don't wish the CSS to be pushed to the client at all then you may set `generateCSS` to false in the global options.

Global Configuration Options
----------------------------
The package exports a global `Avatar` object which has a property named `options` (also an object). The default options provided can be overridden by calling `Avatar.setOptions()` and passing in the options that you wish to override.

***Note that `Avatar.setOptions()` must be called in both server and client code.***

  - `fallbackType`: Determines the type of fallback to use when no image can be found via linked services (Gravatar included):
    - "default image" (the default option, which will show either the image specified by defaultImageUrl, the package's default image, or a Gravatar default image).
      OR
    - "initials" (show the user's initials)
  - `defaultImageUrl`: This will replace the included package default image URL ("packages/utilities_avatar/default.png"). It can be a relative path (e.g. "images/defaultAvatar.png").
  - `gravatarDefault`: Gravatar default option to use (overrides defaultImageUrl option and included package default image URL). Options are available [here](https://secure.gravatar.com/site/implement/images/#default-image).
  - `emailHashProperty`: This property on the user object will be used for retrieving gravatars (useful when user emails are not published).
  - `customImageProperty`:  If you're storing images URLs in a property on the user object, you can specify it here. You can set it to a function that returns the image URL (`this` is the user object).
  - `cssClassPrefix`: This property is used to prefix the CSS classes of the DOM elements generated by this package.
  - `imageSizes`: This property defines the avatar sizes that are available to your project
  - `backgroundColor`: The background color for initials avatars (color name or hex value string or a function). The default is gray (`"#AAA"`). Assign your own background color scheme by passing a function that take only one argument, the `user` instance and return a CSS compatible color code e.g. "#123abc".
  - `textColor`: The text color for initials avatars (color name or hex value string or a function). The default is white (`"#FFF"`). Assign your own text color scheme by passing a function that take only one argument, the `user` instance and return a CSS compatible color code e.g. "#123abc".
  - `generateCSS`: When true the CSS required to style the Avatars based on the above settings is generated and pushed to the client when the application is loaded.

Example usage:
- To show initials when no avatar image can be found via linked services:
```javascript
Avatar.setOptions({
  fallbackType: "initials"
});
```

- To show the included package default image, you don't need to specify any options because this is the default functionality. However, you could specify it explicitly like so:
```javascript
Avatar.setOptions({
  fallbackType: "default image"
});
```
- To show a custom default image:
```javascript
Avatar.setOptions({
  defaultImageUrl: "img/default-avatar.png" OR "http://example.com/default-avatar.png"
});
```
  ***Note that Gravatar's default option requires a publicly accessible URL, so it won't work when your app is running on localhost and you're using either the included package default image or a custom defaultImageUrl that is a relative path. It will work fine once deployed though.***

- To show a custom image property defined on the user object
```javascript
Avatar.setOptions({
  customImageProperty: "imageURL"
});
```
where the value of `imageURL` on the user object may either be a string, or a function that returns the image URL. You may also pass a function directly:

```javascript
Avatar.setOptions({
  customImageProperty: function() {
    var user = this;
    // calculate the image URL here
    return user.profileImageUrl;
  }
});
```

- To show one of Gravatar's default options (e.g. "identicon"):
```javascript
Avatar.setOptions({
  gravatarDefault: "identicon"
});
```
  ***Note that gravatarDefault overrides defaultImageUrl and the included package default image.***

- Set a value to the `cssClassPrefix` to avoid conflicts in CSS class names. Setting the property value like this:
```javascript
Avatar.setOptions({
  cssClassPrefix: 'foo'
});
```
Will assign CSS class names prefixed by 'foo', like below:
```html
<div class="foo foo-small foo-circle">
    <img class="foo-image" src="path_to_your_avatar" alt="avatar" onerror="this.style.display='none';" width="30" height="30">
    <span class="foo-initials">A</span>
</div>
```
If a `cssClassPrefix` is not set, then the default prefix 'avatar' is used for CSS class names in the rendered the HTML:
```html
<div class="avatar avatar-small avatar-circle">
    <img class="avatar-image" src="path_to_your_avatar" alt="avatar" onerror="this.style.display='none';" width="30" height="30">
    <span class="avatar-initials">A</span>
</div>
```

- If your app does not publish the user.emails object/property but publishes an email hash property instead, you can specify it like this (the Gravatar package generates a hash internally when you give it an email too; this just allows you to decouple those two steps so as not to make all your users' emails public):
```javascript
Avatar.setOptions({
  emailHashProperty: "email_hash"
});
```
  ***Note that you can use `Avatar.hash("address@email.com");` to generate a Gravatar email hash.***

- You may override the default avatar sizes, or create new ones, by overriding the `imageSizes` option:
```javascript
Avatar.setOptions({
  imageSizes: {
    'large': 80,
    'mySize': 40
  }  
});
```

- To use a custom coloring scheme for the initials, first write a function that implements your coloring scheme, then assign that to the `backgroundColor` option:
```javascript
var customColorScheme = function (user) {
  if (user.profile.name === "Red")
    return "#ff0000";
  else if (user.profile.name === "Green")
    return "#00ff00";
  else
    return "#aaaaaa";
};

Avatar.setOptions({
  backgroundColor: customColorScheme,
});
```

How the package chooses an avatar
---------------------------------
Given a user object or userId string, Avatar will retrieve the user's image with the following priority:
  1. The user object's custom image URL (stored in the property defined by `Avatar.options.customImageProperty`)
  2. Twitter
  3. Facebook
  4. Google
  5. GitHub
  6. Instagram
  7. Linkedin
  8. Gravatar, which will try to return an avatar matching the user's email address/hash. If it can't find one, then:
    - If `Avatar.options.fallbackType` is "initials", Gravatar returns a 404 (Not Found) response.
    - Else,
      - If `Avatar.options.gravatarDefault` is valid, Gravatar will return a default image (e.g. an identicon).
      - If `Avatar.options.gravatarDefault` is invalid or undefined, Gravatar will return either the image referenced by `Avatar.options.defaultImageUrl` or the included default image.
  9. If no image can be retrieved, the user's initials will be shown.

**Required Fields/Properties on the User Object**

Since fields in `user.services` contain security info, it's often wise to restrict access to those in publications, e.g.:
```javascript
UsersCollection.find({ /* query */ }, {
  fields: {
    //...
    "services.facebook.id" : true
    //...
  }
});
```

Fields used to get avatar image (one per service):
```javascript
"services.twitter.profile_image_url_https"
"services.facebook.id"
"services.google.picture"
"services.github.username"
"services.instagram.profile_picture"
"services.linkedin.pictureUrl"
```

Fields used to form initials (if needed):
```javascript
 "profile.firstName"
 "profile.lastName"
 "profile.familyName"
 "profile.secondName"
 "profile.name"
```

**Linked Services/Accounts:**
By default, the Meteor accounts system creates a separate user account for each service you login with. In order to merge those accounts together, you'd need to use a package like [accounts-meld](https://atmospherejs.com/splendido/accounts-meld) or [link-accounts](https://atmospherejs.com/bozhao/link-accounts).

Credits
-------
- [Sacha Greif](https://github.com/SachaG), for [suggesting the idea on crater.io](http://crater.io/posts/BfMsgzs5AzEdp6Byu)
- [Shai Alon](https://github.com/shaialon), for [contributing the Gravatar functionality to Telescope](https://github.com/TelescopeJS/Telescope/pull/436) that [I later modified](https://github.com/TelescopeJS/Telescope/pull/438)
- [Jérémie Parker](https://github.com/p-j), for providing the [gravatar package](https://github.com/p-j/meteor-gravatar)
- [Everyone who has contributed](https://github.com/meteor-utilities/avatar/graphs/contributors) to this project. :)


History.md
### 0.9.2

- Added option to NOT push the generated CSS into the head
- Use the right URL for the default avatar

### 0.9.1

- Meteor 1.2 compatibility upgrade

### 0.9.0

- React support has had to be removed for the time being. Please see the README for more details
- Added the ability to specify a function that returns the backgroud/foreground colors for a given avatar
- The correct path is now used the default Avatar

### 0.8.0

- Added support for rendering with React (if it's being used by the project)
- Exported a `hash` function to for creating Gravatar-compatible hashes to store in the user object
- Added options to provide the default background/text colors for avatars that display initials

### 0.7.15

- **IMPORTANT:** There is now an `Avatar.setOptions()` method which must be used to override the default options (rather than just assigning to `Avatar.options` directly).
- Provided an option to override the default Avatar sizes and to create new ones.
- The CSS is now generated on the server based on the current options, so it's no longer necessary to provide your own CSS if you've customised the CSS prefix using the option that was introduced in version 0.7.13.


### 0.7.14

- Updated repo URL.

### 0.7.13

- Added custom CSS prefix option (thanks @raiyankamal!)

### 0.7.12

- Cleaned up getServices (thanks @gwendall!)
- Only remove 'display' if url changes (thanks @rodrigok!)

### 0.7.11

- Added LinkedIn support (thanks @raiyankamal!)

### 0.7.10

### 0.7.9

- Adding `customImageProperty` option for storing avatars on a custom user property.
- Refactored `getServices`.
- Started using `getDescendantProp` instead of assuming properties are stored on the `user.profile` object.
- Added roadmap.
- Changed package name to `utilities:avatar`.
- Changed logo to the only “Avatar” that matters.


export.js
// Avatar object to be exported

Avatar = {

  // Default functionality. You can override these options by calling
  // Avater.setOptions (do not set Avatar.options directly)

  options: {

    // Determines the type of fallback to use when no image can be found via
    // linked services (Gravatar included):
    //   "default image" (the default option, which will show either the image
    //   specified by defaultImageUrl, the package's default image, or a Gravatar
    //   default image)
    //     OR
    //   "initials" (show the user's initials).
    fallbackType: '',

    // This will replace the included default avatar image's URL
    // ('packages/utilities_avatar/default.png'). It can be a relative path
    // (relative to website's base URL, e.g. 'images/defaultAvatar.png').
    defaultImageUrl: '',

    // This property name will be used to fetch an avatar url from the user's profile
    // (e.g. 'avatar'). If this property is set and a property of that name exists
    // on the user's profile (e.g. user.profile.avatar) that property will be used
    // as the avatar url.
    customImageProperty: '',

    // Gravatar default option to use (overrides default image URL)
    // Options are available at:
    // https://secure.gravatar.com/site/implement/images/#default-image
    gravatarDefault: '',

    // This property on the user object will be used for retrieving gravatars
    // (useful when user emails are not published).
    emailHashProperty: '',

    // This property is used to prefix the CSS classes of the DOM elements.
    // If no value is set, then the default CSS class assigned to all DOM elements are prefixed with 'avatar' as default.
    // If a value is set to, 'foo' for example, the resulting CSS classes are prefixed with 'foo'.
    cssClassPrefix: '',

    // This property defines the various image sizes available
    imageSizes: {
      'large': 80,
      'small': 30,
      'extra-small': 20
    },

    // Default background color when displaying the initials.
    // Can also be set to a function to map an user object to a background color.
    backgroundColor: "#aaa",

    // Default text color when displaying the initials.
    // Can also be set to a function to map an user object to a text color.
    textColor: "#fff",

    // Generate the required CSS and includ it in the head of your application.
    // Setting this to false will exclude the generated CSS and leave the
    // avatar unstyled by the package.
    generateCSS: true
  },

  // Sets the Avatar options. You must use this setter function rather than assigning directly to
  // Avatar.options, otherwise the stylesheet won't be generated.

  setOptions: function(options) {
    Avatar.options = _.extend(Avatar.options, options);
    if (Avatar.options.generateCSS)
      createCSS();
  },

  // Returns the cssClassPrefix property from options
  getCssClassPrefix: function () {
    return (Avatar.options.cssClassPrefix)? Avatar.options.cssClassPrefix: 'avatar';
  },

  // Returns a background color for initials
  getBackgroundColor: function (user) {
    if (_.isString(Avatar.options.backgroundColor))
      return Avatar.options.backgroundColor;
    else if (_.isFunction(Avatar.options.backgroundColor))
      return Avatar.options.backgroundColor(user);
  },

  // Returns a text color for initials
  getTextColor: function (user) {
    if (_.isString(Avatar.options.textColor))
      return Avatar.options.textColor;
    else if (_.isFunction(Avatar.options.textColor))
      return Avatar.options.textColor(user);
  },

  // Get the initials of the user
  getInitials: function (user) {

    var initials = '';
    var name = '';
    var parts = [];

    if (user && user.profile && user.profile.firstName) {
      initials = user.profile.firstName.charAt(0).toUpperCase();

      if (user.profile.lastName) {
        initials += user.profile.lastName.charAt(0).toUpperCase();
      }
      else if (user.profile.familyName) {
        initials += user.profile.familyName.charAt(0).toUpperCase();
      }
      else if (user.profile.secondName) {
        initials += user.profile.secondName.charAt(0).toUpperCase();
      }
    }
    else {
      if (user && user.profile && user.profile.name) {
        name = user.profile.name;
      }
      else if (user && user.username) {
        name = user.username;
      }

      parts = name.split(' ');
      // Limit getInitials to first and last initial to avoid problems with
      // very long multi-part names (e.g. "Jose Manuel Garcia Galvez")
      initials = _.first(parts).charAt(0).toUpperCase();
      if (parts.length > 1) {
        initials += _.last(parts).charAt(0).toUpperCase();
      }
    }

    return initials;
  },

  // Get the url of the user's avatar
  getUrl: function (user) {

    var url = '';
    var defaultUrl, svc;

    if (user) {
      svc = getService(user);
      if (svc === 'twitter') {
        // use larger image (200x200 is smallest custom option)
        url = user.services.twitter.profile_image_url_https.replace('_normal.', '_200x200.');
      }
      else if (svc === 'facebook') {
        // use larger image (~200x200)
        url = 'https://graph.facebook.com/' + user.services.facebook.id + '/picture/?type=large';
      }
      else if (svc === 'google') {
        url = user.services.google.picture;
      }
      else if (svc === 'github') {
        url = 'https://avatars.githubusercontent.com/' + user.services.github.username + '?s=200';
      }
      else if (svc === 'instagram') {
        url = user.services.instagram.profile_picture;
      }
      else if (svc === 'linkedin') {
        url = user.services.linkedin.pictureUrl;
      }
      else if (svc === "custom") {
        url = getCustomUrl(user);
      }
      else if (svc === 'none') {
        defaultUrl = Avatar.options.defaultImageUrl || '/packages/utilities_avatar/default.png';
        // If it's a relative path (no '//' anywhere), complete the URL
        if (defaultUrl.indexOf('//') === -1) {
          // Strip starting slash if it exists
          if (defaultUrl.charAt(0) === '/') defaultUrl = defaultUrl.slice(1);
          // Then add the relative path to the server's base URL
          defaultUrl = Meteor.absoluteUrl() + defaultUrl;
        }
        url = getGravatarUrl(user, defaultUrl);
      }
    }

    return url;
  },

  // Create a Gravatar-compatible hash for a given email address
  hash: function (email) {
    return Gravatar.hash(email);
  }
}

// Call setOptions to generate the default CSS. This will be replaced if the user calls setOptions in their own code

Avatar.setOptions({});


package.js
Package.describe({
  name: "utilities:avatar",
  summary: "Consolidated user avatar template (twitter, facebook, gravatar, etc.)",
  version: "0.9.2",
  git: "https://github.com/meteor-utilities/avatar"
});

Package.onUse(function(api) {
  api.versionsFrom(['METEOR@0.9.4.1', 'METEOR@1.1.0.1']);
  api.use(['templating', 'reactive-var'], ['client']);
  api.use(['underscore', 'jparker:gravatar@0.3.1'], ['client', 'server']);
  api.use(['meteorhacks:inject-initial@1.0.2'], ['server']);

  // IMPORTANT:
  //
  // React support is being removed from this project for the time being. The latest updates to Meteor and
  // the React package have made the build unstable, and despite my best efforts I haven't been able to
  // create a single solution that works with both apps that use and don't use React.
  //
  // This will be revisited once Meteor 1.2 is out, or else a separate React fork may be considered.
  //
  // To add support for React we can need to weak link to it so that it's not uncessarily included
  // However we have to force the inclusion of the jsx compiler otherwise our jsx files aren't included.
  // Since JSX is only used when compiling and it isn't delivered to clients this isn't an issue, even for
  // those who aren't using React in their project.

  // api.use('jsx@0.1.5');
  // api.use('react@0.1.7', { weak: true });
  // api.use('react-meteor-data@0.1.5');

  api.addFiles(
    [
      'blaze/avatar.html',
      'blaze/avatar.js',
    ],
    ['client']
  );
  api.addFiles(
    [
      // 'react/avatar.jsx',
      'utils.js',
      'helpers.js',
      'export.js'
    ],
    ['client', 'server']
  );

  api.addAssets(
    [
    'default.png'
    ],
    ['client', 'server']
  );

  api.export('Avatar');
});


blaze/avatar.js
Template.avatar.helpers({

  cssClassPrefix: function () {
    return Avatar.getCssClassPrefix();
  },

  size: function () {
    return sizeClass(this);
  },

  shape: function () {
    return shapeClass(this)
  },

  class: function () {
    return customClasses(this);
  },

  imageUrl: function () {
    var user = this.user ? this.user : Meteor.users.findOne(this.userId);
    var url = Avatar.getUrl(user);
    if (url && url.trim() !== '' && Template.instance().firstNode) {
      var img = Template.instance().find('img');
      if (img.src !== url.trim()) {
        img.style.removeProperty('display');
      }
    }
    return url;
  },

  initialsCss: function () {
    var user = this.user ? this.user : Meteor.users.findOne(this.userId);
    var css = '';

    var backgroundColorProperty = 'background-color: ';
    if (this.bgColor) backgroundColorProperty += this.bgColor;
    else backgroundColorProperty += Avatar.getBackgroundColor(user);
    css += backgroundColorProperty + ';';

    var textColorProperty = 'color: ';
    if (this.txtColor) textColorProperty += this.txtColor;
    else textColorProperty += Avatar.getTextColor(user);
    css += textColorProperty + ';';

    return css;
  },

  initialsText: function () {
    var user = this.user ? this.user : Meteor.users.findOne(this.userId);
    return initialsText(user, this);
  }

});


helpers.js
// Get the account service to use for the user's avatar
// Priority: Twitter > Facebook > Google > GitHub > Instagram > Linkedin
getService = function (user) {
  var services = user && user.services || {};
  if (getCustomUrl(user)) { return 'custom'; }
  var service = _.find([['twitter', 'profile_image_url_https'], ['facebook', 'id'], ['google', 'picture'], ['github', 'username'], ['instagram', 'profile_picture'], ['linkedin', 'pictureUrl']], function(s) { return !!services[s[0]] && s[1].length && !!services[s[0]][s[1]]; });
  if(!service)
    return 'none';
  else
    return service[0];
};

getCustomUrl = function (user) {
  var computeUrl = function(prop) {
    if (typeof prop === 'function') {
      prop = prop.call(user);
    }
    if (prop && typeof prop === 'string') {
      return prop;
    }
  }

  var customProp = user && Avatar.options.customImageProperty;
  if (typeof customProp === 'function') {
    return computeUrl(customProp);
  } else if (customProp) {
    return computeUrl(getDescendantProp(user, customProp));
  }
}

getGravatarUrl = function (user, defaultUrl) {
  var gravatarDefault;
  var validGravatars = ['404', 'mm', 'identicon', 'monsterid', 'wavatar', 'retro', 'blank'];

  // Initials are shown when Gravatar returns 404.
  if (Avatar.options.fallbackType !== 'initials') {
    var valid = _.contains(validGravatars, Avatar.options.gravatarDefault);
    gravatarDefault = valid ? Avatar.options.gravatarDefault : defaultUrl;
  }
  else {
    gravatarDefault = '404';
  }

  var options = {
    // NOTE: Gravatar's default option requires a publicly accessible URL,
    // so it won't work when your app is running on localhost and you're
    // using an image with either the standard default image URL or a custom
    // defaultImageUrl that is a relative path (e.g. 'images/defaultAvatar.png').
    default: gravatarDefault,
    size: 200, // use 200x200 like twitter and facebook above (might be useful later)
    secure: true
  };

  var emailOrHash = getEmailOrHash(user);
  return Gravatar.imageUrl(emailOrHash, options);
};

// Get the user's email address or (if the emailHashProperty is defined) hash
getEmailOrHash = function (user) {
  var emailOrHash;
  if (user && Avatar.options.emailHashProperty && !!getDescendantProp(user, Avatar.options.emailHashProperty)) {
    emailOrHash = getDescendantProp(user, Avatar.options.emailHashProperty);
  }
  else if (user && user.emails) {
    var emails = _.pluck(user.emails, 'address');
    emailOrHash = emails[0] || '00000000000000000000000000000000';
  }
  else {
    // If all else fails, return 32 zeros (trash hash, hehe) so that Gravatar
    // has something to build a URL with at least.
    emailOrHash = '00000000000000000000000000000000';
  }
  return emailOrHash;
};

// Returns the size class to use for an avatar
sizeClass = function(context) {
  // Defaults are 'large', 'small', 'extra-small', but user can add new ones
  return Avatar.options.imageSizes[context.size] ? Avatar.getCssClassPrefix() + '-' + context.size : '';
}

// Returns the shape class for an avatar
shapeClass = function (context) {
  var valid = ['rounded', 'circle'];
  return _.contains(valid, context.shape) ? Avatar.getCssClassPrefix() + '-' + context.shape : '';
}

// Returns the custom class(es) for an avatar
customClasses = function (context) {
  return context.class ? context.class : '';
}

// Returns the initials text for an avatar
initialsText = function(user, context) {
  return context.initials || Avatar.getInitials(user);
}

// Creates the dynamically generated CSS file
//
// CSS is dynamically generated so that we can have both a custom class prefix and also allow for custom sizes
createCSS = function () {

  // We only need to do this on the server

  if (!Meteor.isServer)
    return;

  // The base CSS styles

  var p = '.' + Avatar.getCssClassPrefix();
  var a = p + ' ';

  var css =
    p + ' { \n\
      height: 50px; \n\
      width: 50px; \n\
      position: relative; \n\
    } \n' +
    a + p + '-image, \n' +
    a + p + '-initials { \n\
      height: 100%; \n\
      width: 100%; \n\
      position: absolute; \n\
      top: 0px; \n\
      left: 0px; \n\
    } \n' +
    a + p + '-image { \n\
      z-index: 10; \n\
      background-color: #fff; \n\
    } \n' +
    a + p + '-initials { \n\
      display: block; \n\
      background-size: 100% 100%; \n\
      background-color: ' + Avatar.options.backgroundColor + '; \n\
      color: ' + Avatar.options.textColor +'; \n\
      font-size: 25px; \n\
      line-height: 50px; \n\
      font-family: "Helvetica Neue", Helvetica, "Hiragino Sans GB", Arial, sans-serif; \n\
      text-align: center; \n\
      z-index: 1; \n\
    } \n' +
    p + '-rounded ' + p + '-image, \n' +
    p + '-rounded ' + p + '-initials { \n\
      border-radius: 5px; \n\
    } \n'+
    p + '-circle ' + p + '-image, \n' +
    p + '-circle ' + p + '-initials { \n\
      border-radius: 50%; \n\
    } \n' +
    p + '-hide-image ' + p + '-image { \n\
      display: none; \n\
    } \n' +
    p + '-hide-initials ' + p + '-initials { \n\
      display: none; \n\
    } \n\
  ';

  // CSS for each of the defined sizes

  for (sizeName in Avatar.options.imageSizes) {

    var size = Avatar.options.imageSizes[sizeName];

    css = css + p + '-' + sizeName + ' {' +
      'width: ' + size + 'px; ' +
      'min-width: ' + size + 'px; ' +
      'height: ' + size + 'px;' +
    '}\n' +
    p + '-' + sizeName + ' ' + p + '-initials {' +
      'font-size: ' + size / 2 + 'px; ' +
      'line-height: ' + size + 'px;' +
    '}\n';
  }

  // In order to allow for custom sizes and a custom prefix we need to be able to create a style sheet
  // on the fly. To do this cleanly we use the meteor-hacks:inject package to inject the styles directly
  // into the HTML code before it's sent to the client.

  Inject.rawHead('avatar-styles', '<style>' + css + '</style>');
}
